source_code/main_mcu/src/LOGIC/logic_encryption.c

```diff
@@ -252,4 +271,193 @@ void logic_encryption_ctr_decrypt(uint8_t* data, uint8_t* cred_ctr, uint16_t dat
+*   \param  data    data to sign
+*   \param  len     length of data to sign
+*   \param  sig     output signature
+*/
+void logic_encryption_ecc256_sign(uint8_t const *data, int len, uint8_t * sig)
+{
+    size_t result = br_ecdsa_i15_sign_raw(logic_encryption_br_ec_algo, logic_encryption_sha256_ctx.vtable, data, &logic_encryption_fido2_signing_key, sig);
+
+    if (result == 0)
+    {
+        //TODO: Log(TAG_ERR, "error, signing failed!\n");
+        while(1);
+    }
+    //Clear private key used to limit leaking
+    memset(logic_encryption_fido2_priv_key_buf, 0, sizeof(logic_encryption_fido2_priv_key_buf));
+}
+
+/*! \fn     logic_encryption_create_fido2_priv_signing_key(uint8_t const *priv_key, br_ec_private_key *br_key)
+*   \brief  Create key compatible with BearSSL library
+*   \param  priv_key    Buffer containing the private key
+*   \param  br_key      Output BearSSL compatible key
+*/
+static void logic_encryption_create_fido2_priv_signing_key(uint8_t const *priv_key, br_ec_private_key *br_key)
+{
+    memcpy(logic_encryption_fido2_priv_key_buf, priv_key, sizeof(logic_encryption_fido2_priv_key_buf));
+
+    br_key->curve = logic_encryption_br_ec_algo_id;
+    br_key->xlen = FIDO2_PRIV_KEY_LEN;
+    br_key->x = logic_encryption_fido2_priv_key_buf;
+}
+
+/*! \fn     logic_encryption_ecc256_load_key(uint8_t const *key)
+*   \brief  Load a key from a buffer
+*   \param  key     The key to load
+*/
+void logic_encryption_ecc256_load_key(uint8_t const *key)
+{
+    logic_encryption_create_fido2_priv_signing_key(key, &logic_encryption_fido2_signing_key);
+}
+
+/*! \fn     logic_encryption_ecc256_generate_private_key(uint8_t *priv_key)
+*   \brief  Generate a private key using ECC256
+*   \param  priv_key    Output private key
+*   \return size of key
+*/
+size_t logic_encryption_ecc256_generate_private_key(uint8_t *priv_key)
+{
+    size_t result;
+
+    result = br_ec_keygen(&logic_encryption_hmac_drbg_ctx.vtable, logic_encryption_br_ec_algo, NULL, priv_key, logic_encryption_br_ec_algo_id);
+    if (result == 0)
+    {
+        //Log("error, generating private key failed\n");
+        while(1);
+    }
+    return result;
+}
+
+/*! \fn     logic_encryption_ecc256_derive_public_key(uint8_t const *priv_key, ecc256_pub_key *pub_key)
+*   \brief  Derive public key from the private key
+*   \param  priv_key    Private key to derive from
+*   \param  pub_key     Output public key
+*/
+void logic_encryption_ecc256_derive_public_key(uint8_t const *priv_key, ecc256_pub_key *pub_key)
+{
+    #define PUBLIC_KEY_LEN 65
+
+    uint8_t pubkey[PUBLIC_KEY_LEN];
+    size_t result;
+    br_ec_private_key br_priv_key;
+
+    logic_encryption_create_fido2_priv_signing_key(priv_key, &br_priv_key);
+
+    result = br_ec_compute_pub(logic_encryption_br_ec_algo, NULL, pubkey, &br_priv_key);
+    if (result == 0)
+    {
+        //Log("error, deriving public key failed\n");
+        while(1);
+    }
+    memmove(pub_key->x, pubkey + 1, FIDO2_PUB_KEY_X_LEN);
```

Comment:
why the +1 and the 65?

Reply:
there is an extra byte at the beginning of the public key that comes from the BearSSL library. I don't remember off the top of my head now what it is for but its not needed and won't work with FIDO2 key spec.

Reply:
how did you find it out? i can't find that info anywhere :(

Reply:
From debugging and reading online about ECC public key representation.